salt/orch/removal.sls

```diff
@@ -0,0 +1,146 @@
+    - require:
+       - salt: update_pillar
+       - salt: update_grains
+
+update_modules:
+  salt.function:
+    - name: saltutil.sync_modules
+    - tgt: '*'
+    - kwarg:
+        refresh: True
+    - require:
+      - salt: update_mine
+
+{% if target in etcd_members %}
+  # if the node is a member of the etcd cluster:
+  # we must choose another node and promote it to be an etcd master
+  {% set non_etcd_members = salt.saltutil.runner('mine.get', tgt='not G@roles:etcd', fun='network.interfaces', tgt_type='compound').keys() %}
+  {% if non_etcd_members|length > 0 %}
+    {% set new_etcd_member = non_etcd_members[0] %}
+
+    # promote some other node to be a full member of the etcd cluster
+    # WARNING: we should do a consistency check in Velum...
+    #          if no proxy is available for promotion and:
+    #          * num_members >= 3  ->  we should show a "are you sure??" message.
+    #          * num_members == 1  ->  we should just not allow it
+    #
+add_new_etcd_member:
+  salt.state:
+    - tgt: '{{ new_etcd_member }}'
+    - sls:
+      - etcd.remove-pre-stop-services
+    - require:
+      - salt: update_modules
+  {% endif %}
+
+  # remove the etcd master running in {{ target }}
+etc_remove_master:
+  salt.state:
+    - tgt: '{{ target }}'
+    - sls:
+      - etcd.remove-pre-stop-services
+    - require:
+      {%- if non_etcd_members|length > 0 %}
+      - salt: add_new_etcd_member
+      {%- else %}
+      - salt: update_modules
+      {%- endif %}
+
+{% endif %}
+
+# TODO: if {{ target }} is a k8s master, we should promote a worker... but this is
+#       too hard to do right now
+
+update_grains_again:
+  salt.function:
+    - tgt: '*'
+    - name: saltutil.refresh_grains
+    {%- if target in etcd_members %}
+    - require:
+      - salt: etc_remove_master
+    {%- endif %}
+
+stop_services:
+  salt.state:
+    - tgt: {{ target }}
+    - sls:
+      - container-feeder.stop
+      {%- if target in masters %}
+      - kube-apiserver.stop
+      - kube-controller-manager.stop
+      - kube-scheduler.stop
+      {%- else %}
+      - kubelet.stop
+      - kube-proxy.stop
+      {%- endif %}
+      - docker.stop
+      - etcd.stop
+    - require:
+      - salt: update_grains_again
+
+# revoke certificates
+# TODO
+
+# revoke Salt keys
+# TODO
+
+# remove any other configuration in the machines
+# TODO: all the files in /etc/kubernetes/*, /var/lib/kubernetes/* ??
+remove-pre-reboot:
+  salt.state:
+    - tgt: {{ target }}
+    - sls:
+      - container-feeder.remove-pre-reboot
+      {%- if target in masters %}
+      - kube-apiserver.remove-pre-reboot
+      - kube-controller-manager.remove-pre-reboot
+      - kube-scheduler.remove-pre-reboot
+      {%- else %}
+      - kubelet.remove-pre-reboot
+      - kube-proxy.remove-pre-reboot
+      {%- endif %}
+      - docker.remove-pre-reboot
+      - cni.remove-pre-reboot
+      - etcd.remove-pre-reboot
+    - require:
+      - salt: stop_services
+
+# reboot the node
```

Comment:
Should this be shutting down the node rather than rebooting it? The intent is that this machine will be wiped before it can be reused, so shutting down seems better IMO.